README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# wince

<!-- badges: start -->
<!-- badges: end -->

The goal of wince is to …

TODO

-   [ ] make create_nc and write_nc separate, so create is a
    function(dimension, extent = NULL, etc)
-   [ ] convert to RNetCDF
-   [ ] include grid_mapping dummy variable
-   [ ] various fixmes, global atts, good names, integer vs float vs
    byte etc.

## Installation

You can install the development version of wince like so:

``` r
# FILL THIS IN! HOW CAN PEOPLE INSTALL YOUR DEV PACKAGE?
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(wince)
## basic example code
f <- write_nc(volcano, extent = c(0, 87, 0, 61), zvar = -1, z_type = "elevation")
tidync::tidync(f)
#> 
#> Data Source (1): file1d64e43cade63f.nc ...
#> 
#> Grids (4) <dimension family> : <associated variables> 
#> 
#> [1]   D1,D2,D0 : volcano    **ACTIVE GRID** ( 5307  values per variable)
#> [2]   D0       : time
#> [3]   D1       : x
#> [4]   D2       : y
#> 
#> Dimensions 3 (all active): 
#>   
#>   dim   name  length   min   max start count  dmin  dmax unlim coord_dim 
#>   <chr> <chr>  <dbl> <dbl> <dbl> <int> <int> <dbl> <dbl> <lgl> <lgl>     
#> 1 D0    time       1  -1    -1       1     1  -1    -1   TRUE  TRUE      
#> 2 D1    x         87   0.5  86.5     1    87   0.5  86.5 FALSE TRUE      
#> 3 D2    y         61   0.5  60.5     1    61   0.5  60.5 FALSE TRUE

png <- system.file("textures/world.png", package = "rgl", mustWork = TRUE)
arr <- aperm(png::readPNG(png), c(2, 1, 3))
arr <- arr[,ncol(arr):1, ]
f <- write_nc(arr, extent = c(-180, 180, -90, 90), data_name = "world_image")
terra::plotRGB(terra::rast(f) * 256)
maps::map(add = T)
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r


## in future (we need -expand too though for this palette PNG, or nativeRaster etc)
# imvrt <- vapour::vapour_vrt(png, extent = c(-180, 180, -90, 90), projection = "OGC:CRS84", bands = 1:3)
# arr <- whatarelief::imagery(source = imvrt, extent = ex <- c(-1, 1, -1, 1) * 1e7, projection = "+proj=laea")
# write_nc(arr, extent = ex, projection = "+proj=laea")
```

## Code of Conduct

Please note that the wince project is released with a [Contributor Code
of
Conduct](https://contributor-covenant.org/version/2/1/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# wince dev

* Super dodgy new function 'write_var' to just shove arrays in. 

* It's alive. 


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(wince)

test_check("wince")


R/transport.R
#define dimensions
transport_dimensions <- function(bgm, ntime = 365 * 2, nlevels = 2L,
                                 time_unit = "seconds since 1970-01-01 00:00:00") {


  timedim   <- ncdim_def("time", "", seq_len(ntime), unlim = TRUE, create_dimvar = FALSE)
  leveldim  <- ncdim_def("level", "", seq_len(nlevels), create_dimvar = FALSE)
  facesdim  <- ncdim_def("faces", "", seq_len(bgm$extra$nface), create_dimvar = FALSE)

  var.time  <- ncvar_def("time",  time_unit,      dim = timedim,   prec = "double")
  var.face  <- ncvar_def("faces", "",             dim = facesdim,  longname="Face IDs", prec='integer')
  var.lev   <- ncvar_def("level", "",              dim = leveldim,  longname="layer index; 1=near surface", prec="integer")
  var.trans <- ncvar_def("transport", "m3/s",    dim = list(leveldim,facesdim,timedim), 0, prec="float")
  var.destb <- ncvar_def("dest_boxid","id",      dim = facesdim, longname="ID of destination box", prec="integer")
  var.srcb  <- ncvar_def("source_boxid","id",  facesdim,     longname="ID of source box",prec="integer")

  var.pt1x  <- ncvar_def("pt1_x", "degree_east", facesdim, longname = "x-coord of pt 1 of face", prec='float')
  var.pt2x  <- ncvar_def("pt2_x", "degree_east", facesdim, longname = "x-coord of pt 2 of face", prec='float')
  var.pt1y  <- ncvar_def("pt1_y", "degree_north", facesdim, longname = "y-coord of pt 1 of face", prec='float')
  var.pt2y  <- ncvar_def("pt2_y", "degree_north", facesdim, longname = "y-coord of pt 2 of face", prec='float')


  list(var.time, var.face, var.lev, var.destb, var.srcb, var.pt1x, var.pt2x, var.pt1y, var.pt2y, var.trans)
}
#' Create "transport" file for Atlantis.
#'
#' Transport is the hydrodynamic forcing for the Atlantis model from oceanographic models. These functions
#' aim to minimize the amount of manual handling of details, creating an NetCDF file that can be modified directly.
#' All the Box Geometry Model details are determined from the bgm file itself. The output file is the precursor file
#' for input to HydroConstruct (as per section 4.1 of the Atlantis User Guide).
#'
#' Please use the original BGM file in its native map projection, there is no need to copy this to a longitude-latitude version
#' of the same file.
#'
#' TODO: check that bgmlevels should only be an index, check that time_steps must be regular ?( or just monotonic incresing?)
#' @param filename the NetCDF file to create
#' @param model_title name of the model
#' @param bgmfilepath path to BGM file
#' @param bgmlevels the Atlantis depth levels
#' @param time_steps actual time steps, need to be regularly space
#' @param transp_params optional details to put in the NetCDF notes
#' @param overwrite set to \code{TRUE} to clobber an existing file
#' @return the filename of the output (use ncdf4 to inspect, modify it)
#' @export
#' @importFrom ncdf4 nc_open ncatt_put ncvar_put nc_close ncdim_def ncvar_def
#' @examples
#' tfile <- file.path(tempdir(), "my_hydrofile.nc")
#' (ncfile <- create_transport(tfile, bgmfilepath = bgmfiles::bgmfiles("antarctica_99")))
#' ## use ncdf4::nc_open to inspect the file
#'

create_transport <- function(filename,
                             model_title = "Transport file [placeholder]",
                             bgmfilepath = "",
                             bgmlevels = NULL,
                             time_steps = NULL,
                             transp_params = "", overwrite = FALSE) {


  bgm <- rbgm::bgmfile(bgmfilepath)

  if (is.null(bgmlevels)) {
    deltas <- rbgm::build_dz(as.numeric(bgm$extra$maxwcbotz))
    bgmlevels <- cumsum(c(bgm$extra$maxwcbotz, deltas))
    message(sprintf("'bgmlevels' not supplied, building levels at %s", paste(bgmlevels, collapse = ", ")))

  }
  nlevels <- length(bgmlevels)
  if (is.null(time_steps)) {
    time_steps <- seq(as.POSIXct("1970-01-01 00:00:00", tz = "UTC"), by = 12 * 3600, length = 10)
  }

  stopifnot(inherits(time_steps, "POSIXct"))
  stopifnot(length(unique(diff(unclass(time_steps)))) == 1)

  dimensions <- transport_dimensions(bgm, nlevels = nlevels)

  if (file.exists(filename) && !overwrite) {
    stop(sprintf("'filename' already exists, use 'overwrite = TRUE' or delete: \n%s", filename))
  }
  nc_transp <-  ncdf4::nc_create(filename, dimensions)
  on.exit(ncdf4::nc_close(nc_transp), add = TRUE)
  #assign global attributes to file
  ncdf4::ncatt_put(nc_transp, 0, "title",      model_title)
  ncatt_put(nc_transp, 0, "geometry",   basename(bgmfilepath))
  ncatt_put(nc_transp, 0, "parameters", transp_params)
  #assign attributes to variables
  ncatt_put(nc_transp, "time", "dt", diff(time_steps[1:2]), prec="double")
  ncdf4::ncvar_put(nc_transp, "time", time_steps, count = length(time_steps))
  ncvar_put(nc_transp, "faces",  bgm$faces$.fx0)  ## or bgm$faces$.fx0 ??
  ncvar_put(nc_transp, "level",   seq_along(bgmlevels), count = length(bgmlevels))
  ## from RyanM
  # positive flow is right to left across face from p1 to p2
  ncvar_put(nc_transp, "dest_boxid",    bgm$faces$left)
  ncvar_put(nc_transp, "source_boxid",  bgm$faces$right)
  ll <- bgm$vertices
  if (!raster::isLonLat(bgm$extra$projection)) {
    ll[c("x", "y")] <- reproj::reproj(as.matrix( ll[c("x", "y")]), source = bgm$extra$projection,
                                      target = "+proj=longlat +datum=WGS84")[,1:2]
  }
  v1 <- bgm$facesXverts[bgm$facesXverts[[".p0"]] == 1, ] %>%  dplyr::inner_join(ll, ".vx0")
  v2 <- bgm$facesXverts[bgm$facesXverts[[".p0"]] == 2, ] %>%  dplyr::inner_join(ll, ".vx0")


  ncvar_put(nc_transp, "pt1_x",  v1$x)
  ncvar_put(nc_transp, "pt2_x",  v2$x)
  ncvar_put(nc_transp, "pt1_y",  v1$y)
  ncvar_put(nc_transp, "pt2_y",  v2$y)
  filename
}


R/grid_mapping.R

#' Add grid mapping
#'
#' grid_mapping by CF convention
#'
#' mapping is a list with elements 'name' of the crs variable, and 'atts' list
#' with named attributes and values (e.g. standard_parallel = c(10, 20))
#'
#' @param x file name
#' @param overwrite enforce user override to actually update the file
#' @param mapping list of name of variable and attributes for mapping params (see DEtails)
#'
#' @export
#' @return the name of the file modified, returned invisibly
#' @examples
#' mapping <- list(name = "crs",
#'                    atts = list(grid_mapping_name = "lambert_conformal_conic",
#'                                standard_parallel = c(-10, -50),
#'                                #standard_parallel = -50,
#'                                latitude_of_projection_origin = -30,
#'                                longitude_of_central_meridian =  134.33,
#'                                semi_major_axis = 6370000))
#'
add_grid_mapping <- function(x, mapping, overwrite = FALSE) {
  stopifnot(file.exists(x))
  if (!overwrite) stop("file will not be updated unless 'overwrite = TRUE'")
  nc <- try(RNetCDF::open.nc(x, write = TRUE))
  if (inherits(nc, "try-error")) stop()

  var <-   try(RNetCDF::var.def.nc(nc, mapping$name, "NC_INT", NA))
  if (inherits(var, "try-error")) stop(sprintf("\nLooks like file already has variable '%s'?", mapping$name))
  RNetCDF::var.put.nc(nc, mapping$name, 1L)
  #att.put.nc(nc, mapping$name, "_FillValue", "NC_INT", -9L)
   RNetCDF::att.put.nc(nc, mapping$name, "comment", "NC_CHAR",

            "This is a container variable that describes the grid_mapping used by the data in this file. This variable does not contain any data; only information about the geographic coordinate system.")

  nctype <- c(character = "NC_CHAR", double = "NC_FLOAT")
  for (i in seq_along(mapping$atts)) {
    RNetCDF::att.put.nc(nc, mapping$name, names(mapping$atts)[i], nctype[typeof(mapping$atts[[i]])],
                        mapping$atts[[i]])
  }


vars <- ncmeta::nc_vars(nc) |>  dplyr::filter(ndims == 2) |>  dplyr::pull(name)
for (varname in vars) {
  RNetCDF::att.put.nc(nc, varname, "grid_mapping", "NC_CHAR",  mapping$name)
}

RNetCDF::close.nc(nc)
invisible(x)
}


R/write_nc.R


raster_variables <- function(data, extent = NULL,
                             zvar = NULL, z_type = "time",
                             data_name = "some.data", data_unit = "some.unit", long_name = "some.long.name") {
  dimension <- dim(data)



  xx <- vaster::x_centre(dimension, extent)
  yy <- vaster::y_centre(dimension, extent)

  x_dim <-   ncdf4::ncdim_def("x", "metres or degrees WIP", xx, create_dimvar = T)
  y_dim <-   ncdf4::ncdim_def("y", "metres or degrees WIP", yy, create_dimvar = T)



  var.z <- NULL
  if (!is.null(z_type)) {
    if (is.null(zvar)) {
      zvar <- 0
      if (z_type == "time") zvar <- Sys.time()
    }

    z_stuff <- switch(z_type,
                      time = list(unit = "seconds since 1970-01-01 00:00:00",
                                  dimension = "time",
                                  unlim = TRUE),
                      elevation = list(unit = "metres", dimension = "elevation", unlim = FALSE),
                      depth = list(unit = "metres", dimension = "depth", unlim = FALSE)
    )
    z_dim   <- ncdf4::ncdim_def(z_stuff$dimension, "", seq_along(zvar), unlim = z_stuff$unlim, create_dimvar = FALSE)

    ## prec should pivot on integer/float status of z
    var.z <- ncdf4::ncvar_def(z_stuff$dimension, z_stuff$unit, z_dim, prec = "double")



    list_dim <- list(x_dim, y_dim, z_dim)
  } else {
    list_dim <- list(x_dim, y_dim)

  }
  var.data     <- ncdf4::ncvar_def(data_name, data_unit, list_dim, NA,    ## might want NULL here
                                   longname = long_name, prec = "float")

  ## WIP we need grid_mapping and pivot on longlat/projected
  ##  https://github.com/mdsumner/fixoisst/blob/4adef5a28c4d12a6684f2578cfb2ec13a37bfe38/fix-oisst.R#L10-L17
  if (!is.null(var.z)) {
    list(z = var.z,
       data = var.data)
  } else {
    list(data = var.data)
  }
}

#' Create raster file in NetCDF.
#'
#' Rasters need extent, dimension, projection - this writer gives the correct metadata for simple
#' raster inputs with minimal generality and no fuss.
#'
#'  These functions
#' aim to minimize the amount of manual handling of details, creating an NetCDF file that can be modified directly.

#' @param filename the NetCDF file to create
#' @param title name of the model
#' @param zvar actual time steps, need to be regularly space
#' @param transp_params optional details to put in the NetCDF notes
#' @param overwrite set to \code{TRUE} to clobber an existing file
#' @return the filename of the output (use ncdf4 to inspect, modify it)
#' @export
#' @importFrom ncdf4 nc_open ncatt_put ncvar_put nc_close ncdim_def ncvar_def
#' @examples
#' png <- system.file("textures/world.png", package = "rgl", mustWork = TRUE)
#' arr <- aperm(png::readPNG(png), c(2, 1, 3))
#' arr <- arr[,ncol(arr):1, ]
#' f <- write_nc(arr, extent = c(-180, 180, -90, 90), data_name = "world_image")
#' # terra::plotRGB(terra::rast(f) * 256)
#' # maps::map(add = T)
write_nc <- function(data, filename = NULL,
                     extent = NULL,
                     title = "raster",
                     zvar = NULL, z_type = NULL,
                     data_name = NULL, data_unit = "some.unit", long_name = "some.long.name",
                     params = "", overwrite = FALSE) {


    ## we assume raster-oriented input, so we reorient to image()
  dimension <- dim(data)

  #dimension <- dim(data)
  d <- seq_len(length(dim(data)))
  d[1:2] <- d[2:1]
  data <- aperm(data, d)
  if (length(dimension) == 3L) {
    data <- data[,dimension[1L]:1L,]
  } else {
    data <- data[,dimension[1L]:1L ]
  }
  dimension <- dimension[2:1]
  if (is.null(extent)) extent <- c(0, dimension[2L], 0, dimension[1L])
  if (is.null(filename)) filename <- tempfile(fileext = ".nc")
  if (is.null(data_name)) data_name <- deparse1(substitute(data))
  data_name_clean <- gsub("\\s+", "", data_name)
  if (!identical(data_name, data_name_clean)) {
    message(sprintf("variable name taken from '%s' may not be suitable - set 'data_name = <sensible variable name>'", data_name))
    data_name <-data_name_clean

  }
  if (length(dimension) == 3) {
    len <- dimension[3L]
  } else {
    len <- 1
  }
  # if (is.null(zvar)) {
  #
  #   zvar <- Sys.time() + seq(0, by = 1, length.out = len)
  # }
  #

 # stopifnot(length(zvar) == len)
 # if (z_type == "time") stopifnot(inherits(zvar, "POSIXct"))
 # if (length(zvar) > 1) stopifnot(length(unique(diff(unclass(zvar)))) == 1)

  variables <- raster_variables(data, extent = extent, zvar = zvar, z_type = z_type,
                                data_name = data_name, data_unit = "some.unit", long_name = "some.long.name")

  if (file.exists(filename) && !overwrite) {
    stop(sprintf("'filename' already exists, use 'overwrite = TRUE' or delete: \n%s", filename))
  }

  nc_varfile <-  ncdf4::nc_create(filename, variables)
  on.exit(ncdf4::nc_close(nc_varfile), add = TRUE)

  #assign global attributes to file
  ncdf4::ncatt_put(nc_varfile,0,"title", title)
  lenzvar <- NULL
  if (!is.null(variables$z)) {
    ## FIXME
    ncdf4::ncatt_put(nc_varfile, variables$z$name, "dt", 86400, prec="double")
    lenzvar <- length(variables$z$vals)

    ncdf4::ncvar_put(nc_varfile, variables$z, variables$z$vals, count = lenzvar)

  }



  ## note this saves us from 1 or none in the 3rd slot
  ncdf4::ncvar_put(nc_varfile, variables$data, data, count = c(dimension[1:2], lenzvar))
  #browser()
#ncdf4::ncatt_put(nc_varfile, variables$data$name, "coordinates",
#                 paste0(unlist(lapply(variables$data$dim, function(a) a$name)), collapse = " "))
  ## FIXME: pivot on grid_longitude, grid_latitude
  ncdf4::ncatt_put(nc_varfile, "x", "standard_name", "projection_x_coordinate")
  ncdf4::ncatt_put(nc_varfile, "y", "standard_name", "projection_y_coordinate")

  filename
}





write_var <- function(data, filename = NULL,
                     title = "array",
                     zvar = NULL, z_type = NULL,
                     data_name = NULL, data_unit = "some.unit", long_name = "some.long.name",
                     params = "", overwrite = FALSE, append = FALSE) {


  if (is.null(data_name)) data_name <- deparse(substitute(data))
    ## we assume raster-oriented input, so we reorient to image()
  dimension <- dim(data)

  #dimension <- dim(data)
  d <- seq_len(length(dim(data)))
  d[1:2] <- d[2:1]
  data <- aperm(data, d)
  if (length(dimension) == 3L) {
    data <- data[,dimension[1L]:1L,]
  } else {
    data <- data[,dimension[1L]:1L ]
  }
  dimension <- dimension[2:1]
  if (is.null(filename)) filename <- tempfile(fileext = ".nc")

  data_name_clean <- gsub("\\s+", "", data_name)
  if (!identical(data_name, data_name_clean)) {
    message(sprintf("variable name taken from '%s' may not be suitable - set 'data_name = <sensible variable name>'", data_name))
    data_name <-data_name_clean

  }
  if (length(dimension) == 3) {
    len <- dimension[3L]
  } else {
    len <- 1
  }


  if (file.exists(filename) && !overwrite && !append) {
    stop(sprintf("'filename' already exists, use 'overwrite = TRUE', 'append = TRUE' or delete: \n%s", filename))
  }

  if (overwrite && append) {
    stop("if overwrite is TRUE append must be FALSE")
  }

  if (append) {
    nc_varfile <-  RNetCDF::open.nc(filename, write = TRUE)
  } else {
    nc_varfile <-  RNetCDF::create.nc(filename, clobber = overwrite)
  }
  on.exit(RNetCDF::close.nc(nc_varfile), add = TRUE)



  #assign global attributes to file
  if (!append) RNetCDF::att.put.nc(nc_varfile, "NC_GLOBAL", name = "something", type =
                                     "NC_CHAR", "aargh")


  if (!append) {
    dim_ids <- c(RNetCDF::dim.def.nc(nc_varfile, "x", dimension[1]),
               RNetCDF::dim.def.nc(nc_varfile, "y", dimension[2]))
  } else {
    dim_ids <- seq_along(dim(data)) - 1
  }

  var <- RNetCDF::var.def.nc(nc_varfile, data_name, "NC_DOUBLE", dimensions = dim_ids)

  RNetCDF::var.put.nc(nc_varfile, data_name, data)
  filename
}
